README.md
# mysqldumptab

A console tool that dumps MySQL tables as whatever-separated data.

## TL;DR;

`mysqldump --tab` actually calls `SELECT * INTO OUTFILE` on the *server*, so it's not really useful when you need to dump a table from a remote server.

`mysqldumptab` leverages that as it dumps a table in the exact same format as `SELECT INTO OUTFILE` but does that on the client side.

## Dependencies

A compiler.

MySQL client libraries and `build-essential` or `devel`, depending on your OS.

## Installation

Just like any software that comes with `configure` scripts:

``` bash
./bootstrap
./configure --disable-shared --prefix=/usr/local
make install clean
```

## Tweaking

The configure script provides several options for tweaking:


| Option                         | Description                                                                              |
| ------------------------------ | ---------------------------------------------------------------------------------------- |
| `--disable-shared`             | Link the resulting binary against static libraries only. **Recommended**.                |
| `--with-mysql-config`          | Provide an alternate path to `mysql_config`.                                             |
| `--with-buffer-block-size=VAL` | Set the growing output buffer block size to `VAL`. The default value is `0x4000` (16KB). |

## Usage

Some options like `-uphPs` are the same as `mysqldump`.

If you pass `-p` with no argument, you'll get a password prompt.

Options that take characters support hex format - `--fields-escaped-by=0x5c`.

```
Usage:
  mysqldumptab [options] <database> <table>

Options:
  -u, --username                    set mysql user []
  -p, --password                    set mysql password []
  -h, --host                        set mysql host [localhost]
  -P, --port                        set mysql port [3306]
  -s, --socket                      set mysql socket
  -C, --charset                     set connection character set [UTF8]
  -o, --output-file                 write output to file instead of stdout

  -H, --help                        display this help and exit
  -V, --version                     print version information and exit
  -M, --mysql-info                  print mysql client version and exit

  -B, --buffer                      buffer query result set
  -S, --select                      fields to select
  -W, --where                       where clause
  -O, --order                       order by
  -G, --group                       group by
  -A, --having                      having clause

  --fields-terminated-by            terminate output fields with the given string
  --fields-enclosed-by              enclose output fields with the given character
  --fields-optionally-enclosed-by   optionally enclose output fields with the given character
  --fields-escaped-by               escape output fields with the given character
  --lines-terminated-by             terminate lines with the given string
```

## Compatibility

`mysqldumptab` aims to be 100% compatible with the output `SELECT INTO OUTFILE` produces.

If not, please file a [bug](https://github.com/tzvetkoff/mysqldumptab/issues/new).

## Speed

Speed comparisons are rather subjective but when running on the same host, `mysqldumptab` is ~25% faster than `mysqldump` and `SELECT INTO OUTFILE`.

```
# mysqldump -uroot test_db table_with_500k_records > out.sql
real    0m1.810s
user    0m1.608s
sys     0m0.178s

# mysqldump -uroot --tab=out test_db table_with_500k_records
real    0m1.810s
user    0m0.006s
sys     0m0.002s

# mysqldumptab -uroot test_db table_with_500k_records > out.tab
real    0m1.333s
user    0m0.590s
sys     0m0.216s
```

## Bugs

[!@#$](https://github.com/tzvetkoff/mysqldumptab/issues/new)


ChangeLog
Jul 31, 2015 - Version 0.1.2
=======================================================================
* Fixed a very ugly bug with --fields-optionally-enclosed-by

Jul 30, 2015 - Version 0.1.1
=======================================================================
* Support hex when passing characters (--fields-escaped-by=0x5c)

Jun 03, 2015 - Version 0.1.0
=======================================================================
* Refactored build system
* Got rid of gettext
* Refactored the code to some extent
* Added -o option

Mar 23, 2015 - Version 0.0.1
=======================================================================
* Initial release


test/test.sh
#!/bin/bash

DIR=$(cd $(dirname ${BASH_SOURCE[0]}) && pwd)

[[ -n ${MYSQL_USER} ]] && MYSQL_USER="-u${MYSQL_USER}"
[[ -n ${MYSQL_PASS} ]] && MYSQL_PASS="-p${MYSQL_PASS}"
[[ -z ${MYSQL_USER} ]] && MYSQL_USER='-uroot'

if [[ -z ${MDT} ]]; then
	MDT=${DIR}/../src/mysqldumptab

	if [[ ! -x ${MDT} ]]; then
		MDT=`type -p mysqldumptab`

		if [[ ! -x ${MDT} ]]; then
			echo "[!!] \`mysqldumptab' binary not found."
			echo "[!!] If you're running the tests from source, please compile it first."
			echo "[!!] If not, please install \`mysqldumptab' in \$PATH or set the \$MDT environment variable."
			exit 1
		fi
	fi
fi

prepare() {
	rm -f ${DIR}/*.tab

	mysql ${MYSQL_USER} ${MYSQL_PASS} -Bse 'DROP DATABASE IF EXISTS mdt_test'
	mysql ${MYSQL_USER} ${MYSQL_PASS} -Bse 'CREATE DATABASE mdt_test CHARSET=utf8 COLLATE=utf8_general_ci'
	mysql ${MYSQL_USER} ${MYSQL_PASS} mdt_test -Bse "CREATE TABLE test_cases (
		id INTEGER UNSIGNED NOT NULL AUTO_INCREMENT,
		f_null TINYINT(1) UNSIGNED NULL,

		f_int INTEGER NOT NULL,
		f_dec DECIMAL(5, 2) NULL,

		f_date DATE NULL,
		f_datetime DATETIME NULL,
		f_timestamp TIMESTAMP NULL,
		f_time TIME NULL,
		f_year_2 YEAR(2) NULL,
		f_year_4 YEAR(4) NULL,

		f_char CHAR(40) NOT NULL,
		f_varchar VARCHAR(255) NOT NULL,
		f_text TEXT NULL,

		f_enum ENUM('val1', 'val2'),
		f_set SET('set1', 'set2', 'set3'),

		f_geometry GEOMETRY NULL,

		PRIMARY KEY (id)
	)"

	for x in {1..3}; do
		mysql ${MYSQL_USER} ${MYSQL_PASS} mdt_test -Bse "INSERT INTO test_cases VALUES (
			NULL,
			NULL,

			0x5f3759df,
			3.14,

			'9999-12-31',
			'9999-12-31 23:59:59',
			'2038-01-19 03:14:07',
			'-838:59:59',
			42,
			1987,

			'Static CHAR(40)',
			'Variable CHAR(255)',
			'Escaping test: \!@#\$%^&*():;\\'\"\\,.<>/?\`~{}[] => Кирилица, мейт! 鋸',

			'val2',
			'set1,set3',

			POINT(15, 20)
		)"
	done
}

compare() {
	d="${DIR}/${2}"
	m="${d/_d.tab/_m.tab}"
	q="${1}"
	f=""

	shift
	shift

	mysql mdt_test -Bse "${q}"
	${MDT} ${MYSQL_USER} ${MYSQL_PASS} ${@} mdt_test test_cases > ${d}

	cmp ${m} ${d} || {
		echo "${q}"
		echo
		cat ${m}
		echo
		cat ${d}
		exit 1
	}
}

prepare

compare "SELECT * FROM test_cases INTO OUTFILE '${DIR}/dump_1_simple_m.tab'" 'dump_1_simple_d.tab'
compare "SELECT * FROM test_cases INTO OUTFILE '${DIR}/dump_2_escaped_m.tab' FIELDS ESCAPED BY '\!'" 'dump_2_escaped_d.tab' '--fields-escaped-by=!'
compare "SELECT * FROM test_cases INTO OUTFILE '${DIR}/dump_3_optionally_enclosed_m.tab' FIELDS OPTIONALLY ENCLOSED BY '*'" 'dump_3_optionally_enclosed_d.tab' '--fields-optionally-enclosed-by=*'
compare "SELECT * FROM test_cases INTO OUTFILE '${DIR}/dump_4_enclosed_by_m.tab' FIELDS ENCLOSED BY '*'" 'dump_4_enclosed_by_d.tab' '--fields-enclosed-by=*'
compare "SELECT * FROM test_cases INTO OUTFILE '${DIR}/dump_5_terminated_by_m.tab' FIELDS TERMINATED BY 'FOO'" 'dump_5_terminated_by_d.tab' '--fields-terminated-by=FOO'
compare "SELECT * FROM test_cases INTO OUTFILE '${DIR}/dump_6_lines_terminated_by_m.tab' LINES TERMINATED BY 'FOO'" 'dump_6_lines_terminated_by_d.tab' '--lines-terminated-by=FOO'
compare "SELECT * FROM test_cases INTO OUTFILE '${DIR}/dump_7_escaped_by_lines_terminated_by_m.tab' FIELDS ESCAPED BY '\!' LINES TERMINATED BY 'FOO'" 'dump_7_escaped_by_lines_terminated_by_d.tab' '--fields-escaped-by=!' '--lines-terminated-by=FOO'
compare "SELECT * FROM test_cases INTO OUTFILE '${DIR}/dump_8_csv_m.tab' FIELDS TERMINATED BY ',' OPTIONALLY ENCLOSED BY '\"'" 'dump_8_csv_d.tab' '--fields-terminated-by=,' '--fields-optionally-enclosed-by="'

echo 'Tests passed'


tests/run.sh
#!/bin/bash

DIR=$(cd $(dirname ${BASH_SOURCE[0]}) && pwd)

[[ -n ${MYSQL_USER} ]] && MYSQL_USER="-u${MYSQL_USER}"
[[ -n ${MYSQL_PASS} ]] && MYSQL_PASS="-p${MYSQL_PASS}"
[[ -z ${MYSQL_USER} ]] && MYSQL_USER='-uroot'

prepare() {
	rm -f ${DIR}/*.tab

	mysql ${MYSQL_USER} ${MYSQL_PASS} -Bse 'DROP DATABASE IF EXISTS mdt_test'
	mysql ${MYSQL_USER} ${MYSQL_PASS} -Bse 'CREATE DATABASE mdt_test CHARSET=utf8 COLLATE=utf8_general_ci'
	mysql ${MYSQL_USER} ${MYSQL_PASS} mdt_test -Bse "CREATE TABLE test_cases (
		id INTEGER UNSIGNED NOT NULL AUTO_INCREMENT,
		f_null TINYINT(1) UNSIGNED NULL,

		f_int INTEGER NOT NULL,
		f_dec DECIMAL(5, 2) NULL,

		f_date DATE NULL,
		f_datetime DATETIME NULL,
		f_timestamp TIMESTAMP NULL,
		f_time TIME NULL,
		f_year_2 YEAR(2) NULL,
		f_year_4 YEAR(4) NULL,

		f_char CHAR(40) NOT NULL,
		f_varchar VARCHAR(255) NOT NULL,
		f_text TEXT NULL,

		f_enum ENUM('val1', 'val2'),
		f_set SET('set1', 'set2', 'set3'),

		f_geometry GEOMETRY NULL,

		PRIMARY KEY (id)
	)"

	for x in {1..3}; do
		mysql ${MYSQL_USER} ${MYSQL_PASS} mdt_test -Bse "INSERT INTO test_cases VALUES (
			NULL,
			NULL,

			0x5f3759df,
			3.14,

			'9999-12-31',
			'9999-12-31 23:59:59',
			'2038-01-19 03:14:07',
			'-838:59:59',
			42,
			1987,

			'Static CHAR(40)',
			'Variable CHAR(255)',
			'Escaping test: \!@#\$%^&*():;\\'\"\\,.<>/?\`~{}[] => Кирилица, мейт! 鋸',

			'val2',
			'set1,set3',

			POINT(15, 20)
		)"
	done
}

compare() {
	d="${DIR}/${2}"
	m="${d/_d.tab/_m.tab}"
	q="${1}"
	f=""

	shift
	shift

	mysql mdt_test -Bse "${q}"
	${DIR}/../mysqldumptab ${MYSQL_USER} ${MYSQL_PASS} ${@} mdt_test test_cases > ${d}

	cmp ${m} ${d} || {
		echo "${q}"
		echo
		cat ${m}
		echo
		cat ${d}
		exit 1
	}
}

prepare

compare "SELECT * FROM test_cases INTO OUTFILE '${DIR}/dump_1_simple_m.tab'" 'dump_1_simple_d.tab'
compare "SELECT * FROM test_cases INTO OUTFILE '${DIR}/dump_2_escaped_m.tab' FIELDS ESCAPED BY '\!'" 'dump_2_escaped_d.tab' '--fields-escaped-by=!'
compare "SELECT * FROM test_cases INTO OUTFILE '${DIR}/dump_3_optionally_enclosed_m.tab' FIELDS OPTIONALLY ENCLOSED BY '*'" 'dump_3_optionally_enclosed_d.tab' '--fields-optionally-enclosed-by=*'
compare "SELECT * FROM test_cases INTO OUTFILE '${DIR}/dump_4_enclosed_by_m.tab' FIELDS ENCLOSED BY '*'" 'dump_4_enclosed_by_d.tab' '--fields-enclosed-by=*'
compare "SELECT * FROM test_cases INTO OUTFILE '${DIR}/dump_5_terminated_by_m.tab' FIELDS TERMINATED BY 'FOO'" 'dump_5_terminated_by_d.tab' '--fields-terminated-by=FOO'
compare "SELECT * FROM test_cases INTO OUTFILE '${DIR}/dump_6_lines_terminated_by_m.tab' LINES TERMINATED BY 'FOO'" 'dump_6_lines_terminated_by_d.tab' '--lines-terminated-by=FOO'
compare "SELECT * FROM test_cases INTO OUTFILE '${DIR}/dump_7_escaped_by_lines_terminated_by_m.tab' FIELDS ESCAPED BY '\!' LINES TERMINATED BY 'FOO'" 'dump_7_escaped_by_lines_terminated_by_d.tab' '--fields-escaped-by=!' '--lines-terminated-by=FOO'
compare "SELECT * FROM test_cases INTO OUTFILE '${DIR}/dump_8_csv_m.tab' FIELDS TERMINATED BY ',' OPTIONALLY ENCLOSED BY '\"'" 'dump_8_csv_d.tab' '--fields-terminated-by=,' '--fields-optionally-enclosed-by="'

echo 'Tests passed'


mdt_str.c
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <math.h>

#include "mdt_str.h"

struct mdt_str_t *
mdt_str_alloc()
{
    struct mdt_str_t *result = calloc(1, sizeof(struct mdt_str_t));

    if (!result) {
        fprintf(stderr, "Cannot allocate dynamic string\n");
        exit(EXIT_FAILURE);
    }

    result->len = 0;
    result->siz = 0x400;
    result->ptr = calloc(0x400 + 1, 1);

    if (!result->ptr) {
        free(result);
        fprintf(stderr, "Cannot allocate dynamic string\n");
        exit(EXIT_FAILURE);
    }

    return result;
}

void
mdt_str_free(struct mdt_str_t *mdt_str)
{
    free(mdt_str->ptr);
    free(mdt_str);
}

void
mdt_str_clear(struct mdt_str_t *mdt_str)
{
    *mdt_str->ptr = '\0';
    mdt_str->len = 0;
}

void
mdt_str_grow(struct mdt_str_t *mdt_str, int amount)
{
    char *ptr;

    ptr = realloc(mdt_str->ptr, mdt_str->siz + amount);
    if (!ptr) {
        fprintf(stderr, "Cannot resize dynamic string to %d\n", mdt_str->siz +
            amount);
        mdt_str_free(mdt_str);
        exit(EXIT_FAILURE);
    }

    mdt_str->siz += amount;
    mdt_str->ptr = ptr;
}

void
mdt_str_concat(struct mdt_str_t *mdt_str, const char *fmt, ...)
{
    va_list ap;
    int len, amount;

    /* estimate size */
    va_start(ap, fmt);
    len = vsnprintf(NULL, 0, fmt, ap);
    va_end(ap);

    /* check if we need to grow the string */
    if (mdt_str->len + len > mdt_str->siz) {
        amount = floor((float)(mdt_str->len + len - mdt_str->siz) /
            (float)0x400) * 0x400;
        mdt_str_grow(mdt_str, amount);
    }

    va_start(ap, fmt);
    vsnprintf(mdt_str->ptr + mdt_str->len, mdt_str->siz - mdt_str->len, fmt, ap);
    mdt_str->len += len;
    va_end(ap);
}


src/str.c
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <math.h>

#include "str.h"

struct mdt_str_t *
mdt_str_alloc()
{
    struct mdt_str_t *result = calloc(1, sizeof(struct mdt_str_t));

    if (!result) {
        fprintf(stderr, "Cannot allocate dynamic string\n");
        exit(EXIT_FAILURE);
    }

    result->len = 0;
    result->siz = 0x400;
    result->ptr = calloc(0x400 + 1, 1);

    if (!result->ptr) {
        free(result);
        fprintf(stderr, "Cannot allocate dynamic string\n");
        exit(EXIT_FAILURE);
    }

    return result;
}

void
mdt_str_free(struct mdt_str_t *mdt_str)
{
    free(mdt_str->ptr);
    free(mdt_str);
}

void
mdt_str_clear(struct mdt_str_t *mdt_str)
{
    *mdt_str->ptr = '\0';
    mdt_str->len = 0;
}

void
mdt_str_grow(struct mdt_str_t *mdt_str, int amount)
{
    char *ptr;

    ptr = realloc(mdt_str->ptr, mdt_str->siz + amount);
    if (!ptr) {
        fprintf(stderr, "Cannot resize dynamic string to %d\n", mdt_str->siz +
            amount);
        mdt_str_free(mdt_str);
        exit(EXIT_FAILURE);
    }

    mdt_str->siz += amount;
    mdt_str->ptr = ptr;
}

void
mdt_str_concat(struct mdt_str_t *mdt_str, const char *fmt, ...)
{
    va_list ap;
    int len, amount;

    /* estimate size */
    va_start(ap, fmt);
    len = vsnprintf(NULL, 0, fmt, ap);
    va_end(ap);

    /* check if we need to grow the string */
    if (mdt_str->len + len > mdt_str->siz) {
        amount = floor((float)(mdt_str->len + len - mdt_str->siz) /
            (float)0x400) * 0x400;
        mdt_str_grow(mdt_str, amount);
    }

    va_start(ap, fmt);
    vsnprintf(mdt_str->ptr + mdt_str->len, mdt_str->siz - mdt_str->len, fmt, ap);
    mdt_str->len += len;
    va_end(ap);
}


src/main.c
#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <getopt.h>

#include <mysql.h>

#include "str.h"

/* default connection stuff */
#define MYSQLDUMPTAB_DEFAULT_USERNAME ""
#define MYSQLDUMPTAB_DEFAULT_PASSWORD ""
#define MYSQLDUMPTAB_DEFAULT_HOST "localhost"
#define MYSQLDUMPTAB_DEFAULT_PORT 3306
#define MYSQLDUMPTAB_DEFAULT_CHARSET "UTF8"

/* define some types that may be missing on older mysql versions */
#ifndef MYSQL_TYPE_TIMESTAMP2
#define MYSQL_TYPE_TIMESTAMP2 MYSQL_TYPE_VARCHAR
#endif

#ifndef MYSQL_TYPE_DATETIME2
#define MYSQL_TYPE_DATETIME2 MYSQL_TYPE_VARCHAR
#endif

#ifndef MYSQL_TYPE_TIME2
#define MYSQL_TYPE_TIME2 MYSQL_TYPE_VARCHAR
#endif

/* print block */
#ifndef MYSQLDUMPTAB_BUFFER_BLOCK_SIZE
#define MYSQLDUMPTAB_BUFFER_BLOCK_SIZE 0x4000
#endif

/* does anyone use this type? */
#define MDT_NULL(x) (\
    (x)->type == MYSQL_TYPE_NULL \
)

/* a nice macro to test if a field should be enclosed */
#define MDT_ENCLOSE(x) (\
    (x)->type == MYSQL_TYPE_VARCHAR || \
    (x)->type == MYSQL_TYPE_DATETIME || \
    (x)->type == MYSQL_TYPE_DATETIME2 || \
    (x)->type == MYSQL_TYPE_DATE || \
    (x)->type == MYSQL_TYPE_NEWDATE || \
    (x)->type == MYSQL_TYPE_TIME || \
    (x)->type == MYSQL_TYPE_TIME2 || \
    (x)->type == MYSQL_TYPE_TIMESTAMP || \
    (x)->type == MYSQL_TYPE_TIMESTAMP2 || \
    ((x)->type >= MYSQL_TYPE_ENUM && (x)->type <= MYSQL_TYPE_GEOMETRY) \
)

/* global variables */
static char *username = NULL, *password = NULL, *host = NULL,
    *sock = NULL, *charset = NULL, *database = NULL, *table = NULL,
    *select_fields = NULL, *where = NULL, *group = NULL,
    *having = NULL, *order = NULL,
    *fields_terminated_by = NULL, fields_enclosed_by = 0,
    fields_optionally_enclosed_by = 0, fields_escaped_by = '\\',
    *lines_terminated_by = NULL, *print_buffer = NULL;
static unsigned int port, fields_terminated_by_length;
static unsigned long print_buffer_size;
static char *enclose;

static FILE *fo = NULL;

static MYSQL *connection = NULL;
static MYSQL_RES *result = NULL;
static MYSQL_FIELD **fields = NULL;

/* exit handler - release memory, close connections, etc. */
static void
exit_handler()
{
    if (username) {
        free(username);
    }
    if (password) {
        free(password);
    }
    if (host) {
        free(host);
    }
    if (sock) {
        free(sock);
    }
    if (charset) {
        free(charset);
    }
    if (database) {
        free(database);
    }
    if (table) {
        free(table);
    }
    if (select_fields) {
        free(select_fields);
    }
    if (where) {
        free(where);
    }
    if (group) {
        free(group);
    }
    if (having) {
        free(having);
    }
    if (order) {
        free(order);
    }
    if (fields_terminated_by) {
        free(fields_terminated_by);
    }
    if (lines_terminated_by) {
        free(lines_terminated_by);
    }
    if (enclose) {
        free(enclose);
    }
    if (print_buffer) {
        free(print_buffer);
    }
    if (fields) {
        free(fields);
    }
    if (fo && fo != stdout) {
        fclose(fo);
    }
    if (result) {
        mysql_free_result(result);
    }
    if (connection) {
        mysql_close(connection);
    }
}

/* usage */
static void
usage(const char *program_invocation_name, FILE *out)
{
    fputs("\n", out);

    fputs("Usage:\n", out);
    fprintf(out, "  %s [options] <database> <table>\n",
        program_invocation_name);
    fputs("\n", out);

    fputs("Options:\n", out);
    fputs("  -u, --username                    set mysql user ["
        MYSQLDUMPTAB_DEFAULT_USERNAME "]\n", out);
    fputs("  -p, --password                    set mysql password ["
        MYSQLDUMPTAB_DEFAULT_PASSWORD "]\n", out);
    fputs("  -h, --host                        set mysql host ["
        MYSQLDUMPTAB_DEFAULT_HOST "]\n", out);
    fprintf(out, "  -P, --port                        set mysql port [%d]\n",
        MYSQLDUMPTAB_DEFAULT_PORT);
    fputs("  -s, --socket                      set mysql socket\n", out);
    fputs("  -C, --charset                     set connection character set ["
        MYSQLDUMPTAB_DEFAULT_CHARSET "]\n", out);
    fputs("  -o, --output-file                 write output to file instead of"
        " stdout\n", out);
    fputs("\n", out);

    fputs("  -H, --help                        display this help and exit\n",
        out);
    fputs("  -V, --version                     print version information "
        "and exit\n", out);
    fputs("  -M, --mysql-info                  print mysql client version "
        "and exit\n", out);
    fputs("\n", out);

    fputs("  -B, --buffer                      buffer query result set\n",
        out);
    fputs("  -S, --select                      fields to select\n", out);
    fputs("  -W, --where                       where clause\n", out);
    fputs("  -O, --order                       order by\n", out);
    fputs("  -G, --group                       group by\n", out);
    fputs("  -A, --having                      having clause\n", out);
    fputs("\n", out);

    fputs("  --fields-terminated-by            terminate output fields "
        "with the given string\n", out);
    fputs("  --fields-enclosed-by              enclose output fields "
        "with the given character\n", out);
    fputs("  --fields-optionally-enclosed-by   optionally enclose output "
        "fields with the given character\n", out);
    fputs("  --fields-escaped-by               escape output fields with the "
        "given character\n", out);
    fputs("  --lines-terminated-by             terminate lines with the "
        "given string\n", out);
    fputs("\n", out);

    exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

/* lazy character parsing */
static char
parse_char(char *arg)
{
    if (arg[0] == '0' && (arg[1] == 'x' || arg[1] == 'X')) {
        return strtoul(arg, NULL, 16) & 0xFF;
    }

    return arg[0];
}

/* entry point */
int
main(int argc, char **argv)
{
    static int ch, n, n1, password_prompt = 0, buffer = 0, enclose_char;
    static unsigned long i, j, k, size_needed, length, *lengths;
    static char *p, *src, *dst, *fo_name = NULL;
    static struct mdt_str_t *query;
    static MYSQL_ROW row;

    static const struct option longopts[] = {
        {"username", required_argument, NULL, 'u'},
        {"password", optional_argument, NULL, 'p'},
        {"host", required_argument, NULL, 'h'},
        {"port", required_argument, NULL, 'P'},
        {"socket", required_argument, NULL, 's'},
        {"charset", required_argument, NULL, 'C'},

        {"help", no_argument, NULL, 'H'},
        {"version", no_argument, NULL, 'V'},
        {"mysql-info", no_argument, NULL, 'M'},

        {"buffer", no_argument, NULL, 'B'},
        {"select", required_argument, NULL, 'S'},
        {"where", required_argument, NULL, 'W'},
        {"order", required_argument, NULL, 'O'},
        {"group", required_argument, NULL, 'G'},
        {"having", required_argument, NULL, 'A'},

        {"fields-terminated-by", required_argument, NULL, 0x1000},
        {"fields-enclosed-by", required_argument, NULL, 0x1001},
        {"fields-optionally-enclosed-by", required_argument, NULL, 0x1002},
        {"fields-escaped-by", required_argument, NULL, 0x1003},
        {"lines-terminated-by", required_argument, NULL, 0x1004},
        {NULL, 0, NULL, 0}
    };

    /* initialize default options */
    username = strdup(MYSQLDUMPTAB_DEFAULT_USERNAME);
    password = strdup(MYSQLDUMPTAB_DEFAULT_PASSWORD);
    host = strdup(MYSQLDUMPTAB_DEFAULT_HOST);
    port = MYSQLDUMPTAB_DEFAULT_PORT;
    sock = NULL;
    charset = strdup(MYSQLDUMPTAB_DEFAULT_CHARSET);
    fo = stdout;

    fields_terminated_by = strdup("\t");
    lines_terminated_by = strdup("\n");
    print_buffer_size = MYSQLDUMPTAB_BUFFER_BLOCK_SIZE;
    print_buffer = malloc(print_buffer_size);

    /* register the exit handler */
    atexit(exit_handler);

    /* parse command line options */
    while ((ch = getopt_long(argc, argv, "u:p::h:P:s:C:o:HVMBS:W:O:G:A:",
        longopts, NULL)) != -1) {
        switch (ch) {
            case 'u':
                free(username);
                username = strdup(optarg);
                break;
            case 'p':
                free(password);
                if (optarg) {
                    password = strdup(optarg);
                } else {
                    password_prompt = 1;
                }
                break;
            case 'h':
                free(host);
                host = strdup(optarg);
                break;
            case 'P':
                port = strtol(optarg, &p, 10);
                if (p == optarg) {
                    fprintf(stderr, "Invalid port: %s\n", optarg);
                    return EXIT_FAILURE;
                }
                break;
            case 's':
                sock = strdup(optarg);
                break;
            case 'C':
                free(charset);
                charset = strdup(optarg);
                break;
            case 'o':
                if (strcmp(optarg, "-") == 0) {
                    fo_name = NULL;
                } else {
                    fo_name = optarg;
                }
                break;

            case 'H':
                usage(argv[0], stdout);
                break;
            case 'V':
                puts(MYSQLDUMPTAB_VERSION_STRING);
                return EXIT_SUCCESS;
            case 'M':
                puts(mysql_get_client_info());
                return EXIT_SUCCESS;

            case 'B':
                buffer = 1;
                break;
            case 'S':
                select_fields = strdup(optarg);
                break;
            case 'W':
                where = strdup(optarg);
                break;
            case 'O':
                order = strdup(optarg);
                break;
            case 'G':
                group = strdup(optarg);
                break;
            case 'A':
                having = strdup(optarg);
                break;

            case 0x1000:
                free(fields_terminated_by);
                fields_terminated_by = strdup(optarg);
                break;
            case 0x1001:
                fields_enclosed_by = parse_char(optarg);
                break;
            case 0x1002:
                fields_optionally_enclosed_by = parse_char(optarg);
                break;
            case 0x1003:
                fields_escaped_by = parse_char(optarg);
                break;
            case 0x1004:
                free(lines_terminated_by);
                lines_terminated_by = strdup(optarg);
                break;

            default:
                usage(argv[0], stderr);
        }
    }

    /* check for 2 positional arguments */
    if (argc - optind != 2) {
        usage(argv[0], stderr);
    }

    /* open output file */
    if (fo_name) {
        if ((fo = fopen(fo_name, "w")) == NULL) {
            fprintf(stderr, "Unable to open output file \"%s\"\n", fo_name);
            return EXIT_FAILURE;
        }
    }

    fields_terminated_by_length = strlen(fields_terminated_by);

    if (password_prompt) {
        p = getpass("Enter password: ");
        password = strdup(p);
        memset(p, 0, strlen(password));
    }

    database = strdup(argv[optind]);
    table = strdup(argv[optind + 1]);

    /* initialize connection */
    if ((connection = mysql_init(NULL)) == NULL) {
        fprintf(stderr, "%s\n", mysql_error(connection));
        return EXIT_FAILURE;
    }

    /* connect to database */
    if (mysql_real_connect(connection, host, username,
        password, database, port, sock, 0) == NULL) {
        fprintf(stderr, "%s\n", mysql_error(connection));
        return EXIT_FAILURE;
    }

    /* just in case */
    free(password);
    password = NULL;

    /* allocate query buffer */
    query = mdt_str_alloc();

    /* set charset */
    if (charset && strlen(charset)) {
        mdt_str_concat(query, "SET CHARSET %s", charset);

        if (mysql_query(connection, query->ptr) != 0) {
            mdt_str_free(query);
            fprintf(stderr, "%s\n", mysql_error(connection));
            return EXIT_FAILURE;
        }
    }
    mdt_str_clear(query);

    /* build query */
    mdt_str_concat(query, "SELECT /*!40001 SQL_NO_CACHE */ %s FROM ",
        select_fields ? select_fields : "*");
    mdt_str_concat(query, "%s", table);

    if (where) {
        mdt_str_concat(query, " WHERE %s", where);
    }
    if (group) {
        mdt_str_concat(query, " GROUP BY %s", group);
    }
    if (having) {
        mdt_str_concat(query, " HAVING %s", having);
    }
    if (order) {
        mdt_str_concat(query, " ORDER BY %s", order);
    }

    /* execute query */
    if (mysql_query(connection, query->ptr) != 0) {
        mdt_str_free(query);
        fprintf(stderr, "%s\n", mysql_error(connection));
        return EXIT_FAILURE;
    }
    mdt_str_free(query);

    /* store || use result */
    result = buffer ? mysql_store_result(connection) :
        mysql_use_result(connection);
    if (!result) {
        fprintf(stderr, "%s\n", mysql_error(connection));
        return EXIT_FAILURE;
    }

    /* fields count */
    n = mysql_num_fields(result);
    n1 = n - 1;

    /* fetch fields information */
    fields = calloc(n, sizeof(struct st_mysql_field));
    enclose = calloc(n, sizeof(char));
    if (!fields || !enclose) {
        fputs("Cannot allocate memory for fields information\n", stderr);
        return EXIT_FAILURE;
    }

    /* enclose char */
    enclose_char = fields_enclosed_by ? fields_enclosed_by :
        fields_optionally_enclosed_by;

    /* figure out whether we should enclose a field based on its type */
    for (i = 0; i < n; ++i) {
        if ((fields[i] = mysql_fetch_field(result)) == NULL) {
            fprintf(stderr, "myslq_field() failed for field #%lu\n", i);
            return EXIT_FAILURE;
        }

        if (fields_enclosed_by) {
            enclose[i] = fields_enclosed_by;
        } else if (fields_optionally_enclosed_by) {
            enclose[i] = MDT_ENCLOSE(fields[i]) ?
                fields_optionally_enclosed_by : 0;
        }
    }

    /* iterate over rows and print data */
    while ((row = mysql_fetch_row(result))) {
        lengths = mysql_fetch_lengths(result);

        size_needed = 1;
        for (j = 0; j < n; ++j) {
            size_needed += lengths[j] * 2 + fields_terminated_by_length;
        }

        if (print_buffer_size < size_needed) {
            print_buffer_size = floor((double)size_needed /
                (double)MYSQLDUMPTAB_BUFFER_BLOCK_SIZE) *
                MYSQLDUMPTAB_BUFFER_BLOCK_SIZE;
            print_buffer = realloc(print_buffer, print_buffer_size);
            if (!print_buffer) {
                fprintf(stderr, "Cannot grow print buffer to %lu bytes\n",
                    print_buffer_size);
                return EXIT_FAILURE;
            }
        }

        dst = print_buffer;

        for (i = 0; i < n; ++i) {
            src = row[i];
            length = lengths[i];

            if (!src || MDT_NULL(fields[i])) {
                *dst++ = fields_escaped_by;
                *dst++ = 'N';
            } else {
                if (enclose[i]) {
                    *dst++ = enclose_char;
                }

                /* walk and escape what we need to escape */
                for (j = 0; j < length; ++j) {
                    ch = *src++;

                    switch (ch) {
                        case '\0':
                            *dst++ = fields_escaped_by;
                            *dst++ = '0';
                            break;

                        case '\t':
                        case '\n':
                            *dst++ = fields_escaped_by;
                            *dst++ = ch;
                            break;

                        default:
                            if (ch == fields_escaped_by ||
                                ch == enclose_char) {
                                *dst++ = fields_escaped_by;
                            }

                            *dst++ = ch;
                    }
                }

                if (enclose[i]) {
                    *dst++ = enclose_char;
                }
            }

            if (i < n1) {
                strncpy(dst, fields_terminated_by,
                    fields_terminated_by_length);
                dst += fields_terminated_by_length;
            }
        }

        *dst = '\0';

        fputs(print_buffer, fo);
        fputs(lines_terminated_by, fo);
    }

    /* hasta la vista, baby! */
    return EXIT_SUCCESS;
}
